j2cl-tasks/src/main/java/com/vertispan/j2cl/build/provided/TranslationsFileProcessor.java

```diff
@@ -97,17 +98,109 @@ public Optional<File> getTranslationsFile(List<Input> inputs, BuildLog log) {
+
+            if (locale.contains("_")) {
+                // according to spec, locale discaration can have 3 parts max
+                if(locale.indexOf("_") != locale.lastIndexOf("_")) {
+                    result.add(locale.substring(0, locale.lastIndexOf("_")));
+                }
+                result.add(locale.substring(0, locale.indexOf("_")));
+            }
+            return result;
+        }
+
+        private Optional<File> mergeFiles(HashMap<String, Set<NodeList>> suitableFiles, Set<String> locales, TaskContext context) {
+            Map<String, Node> resultedCodeSet = new HashMap<>();
+            // parent nodes first
+            locales.stream().sorted((o1, o2) -> {
+                if(o1.split("_").length > o2.split("_").length) {
+                    return 1;
+                }
+                return -1;
+            }).forEach(locale -> {
+                suitableFiles.get(locale).forEach(node -> {
+                    NodeList children = node.item(0).getChildNodes();
+
+                    for (int i = 0; i < children.getLength(); i++) {
+                        if (children.item(i).getNodeName().equals("translation")) {
+                            String key = children.item(i).getAttributes().getNamedItem("id").getNodeValue();
+                            // override parent keys if already exists
+                            resultedCodeSet.put(key, children.item(i));
+                        }
+                    }
+                });
+            });
+
+            Path folder = context.outputPath();
+            File generated = folder.resolve("generated_messages.xtb").toFile();
+            generateAndWriteXTB(resultedCodeSet, generated);
+            return Optional.of(generated);
+        }
+
+        private void generateAndWriteXTB(Map<String, Node> resultedCodeSet, File generated) {
+            StringBuffer sb = new StringBuffer();
+
+            sb.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>");
+            sb.append(System.lineSeparator());
+            sb.append("<!DOCTYPE translationbundle SYSTEM \"translationbundle.dtd\">");
+            sb.append(System.lineSeparator());
+
+            sb.append("<translationbundle lang=\"" + locale + "\">");
+            sb.append(System.lineSeparator());
+
+            resultedCodeSet.values().forEach(node -> {
+                sb.append("  <translation id=\"");
+                sb.append(node.getAttributes().getNamedItem("id").getNodeValue());
+                sb.append("\" key=\"");
+                sb.append(node.getAttributes().getNamedItem("key").getNodeValue());
+                sb.append("\">");
+
+                // we have to process placeholders here
+                for (int i = 0; i < node.getChildNodes().getLength(); i++) {
+                    Node innerTextNode = node.getChildNodes().item(i);
+                    if(innerTextNode.getNodeType() == Node.TEXT_NODE) {
+                        sb.append(innerTextNode.getTextContent());
```

Comment:
These need to be escaped, right? else you could open a new tag in there?

Reply:
in this case, closure ll fail, but looks like your comment is not very clear for me, could you be more specific please ?

Reply:
My concern is that a user has a `<` token in their translation string, expected to be left as a literal or as html (doesnt matter), but instead of breaks the newly created DOM that you are appending things to. XML injection, but not deliberately to break things.